{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "\n",
    "# taking genes from file\n",
    "source = 'C:/AGlab/IB/opened/'\n",
    "final = 'C:/AGlab/gene_sorted.txt'\n",
    "\n",
    "g = open(final, 'w')\n",
    "g.close()\n",
    "\n",
    "files = os.listdir(source)\n",
    "for gff in files:\n",
    "    with open(source + gff, 'r') as f:\n",
    "        string = ''\n",
    "        string += gff.split('_')[2] + ' '\n",
    "        #string += gff.split('.')[0] + ' '\n",
    "        for lines in f:\n",
    "            if not lines.startswith('#'):\n",
    "                if lines.split()[2] == 'gene':\n",
    "                    try:\n",
    "                        if lines.split()[8].split(';')[3].split('=')[1] == 'protein_coding' or \\\n",
    "                            lines.split()[8].split(';')[3].split('=')[1] == 'Gene':\n",
    "                            None\n",
    "                        else:\n",
    "                            #string += lines.split()[8].split(';')[0].split('-')[1].split('_')[1] + ' '\n",
    "                            string += lines.split()[8].split(';')[3].split('=')[1] + ' '\n",
    "                    except IndexError:\n",
    "                        None\n",
    "        string += '\\n'\n",
    "        g = open(final, 'a')\n",
    "        if len(string.split()) > 1000:\n",
    "            g.write(string)\n",
    "        #print(gff.split('_')[2])\n",
    "        g.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "\n",
    "# making table for R vizualization \n",
    "file = 'C:/AGlab/gene_sorted.txt'\n",
    "\n",
    "gene_names = {}\n",
    "assmeblies_names = []\n",
    "with open(file, 'r') as f:\n",
    "    for lines in f:\n",
    "        main_list = lines.split()\n",
    "        n = main_list[0]\n",
    "        assmeblies_names.append(n)\n",
    "        main_list.pop(0)\n",
    "        for genes in main_list:\n",
    "            if genes in gene_names.keys():\n",
    "                gene_names[genes].append(n)\n",
    "            else:\n",
    "                gene_names[genes] = [n]\n",
    "\n",
    "# for i, v in gene_names.items():\n",
    "#     if set(v) == 5:\n",
    "#         print(i, set(v))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "\n",
    "# making statistics for counting filetic patterns\n",
    "dna = []\n",
    "\n",
    "source = 'C:/AGlab/IB/opened/'\n",
    "final = 'C:/AGlab/gene_around.txt'\n",
    "ended = 'C:/AGlab/around.txt'\n",
    "\n",
    "g = open(final, 'w')\n",
    "g.write('name ' + 'start ' + 'end ' + 'strand' + '\\n')\n",
    "g.close()\n",
    "\n",
    "files = os.listdir(source)\n",
    "for gff in files:\n",
    "    for var in dna:\n",
    "        if var in gff:\n",
    "            with open(source + gff, 'r') as f:\n",
    "                with open(final, 'a') as d:\n",
    "                    for lines in f:\n",
    "                        if not lines.startswith('#'):\n",
    "                            if lines.split()[2] == 'gene':\n",
    "                                d.write(lines.split()[8].split(';')[1].split('=')[1] +\n",
    "                                        ' ' +\n",
    "                                        lines.split()[3] + ' ' +\n",
    "                                        lines.split()[4] + ' ' +\n",
    "                                        lines.split()[6] + ' ' +\n",
    "                                        str(var) +\n",
    "                                        '\\n')\n",
    "\n",
    "with open(final, 'r') as f:\n",
    "    with open(ended, 'w') as e:\n",
    "        n = 0\n",
    "        for lines in f:\n",
    "            if lines.split()[0] == 'prkT':\n",
    "                m = 0\n",
    "                with open(final, 'r') as f1:\n",
    "                    e.write(lines.split()[4] + '\\n')\n",
    "                    for lines in f1:\n",
    "                        if n - 10 < m < n + 10:\n",
    "                            e.write(lines.split()[0] + ' ' +\n",
    "                                    lines.split()[1] + ' ' +\n",
    "                                    lines.split()[2] + ' ' +\n",
    "                                    lines.split()[3] + '\\n')\n",
    "                        m += 1\n",
    "                e.write('\\n')\n",
    "            n += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "\n",
    "# script for working with 16S\n",
    "path = r'C:\\Users\\ALEX\\Desktop\\assemblies'\n",
    "\n",
    "def get_16s(seq, value):\n",
    "    result = []\n",
    "    for rec in SeqIO.parse(seq, 'genbank'):\n",
    "        for feature in rec.features:\n",
    "            if value in feature.qualifiers.get('product', []):\n",
    "                result.append(feature.location.extract(rec.seq))\n",
    "    return result\n",
    "\n",
    "# rna_names = get_16s_names(os.path.join(path, assembly))\n",
    "rrna = ['ribosomal RNA-16S', '16S ribosomal RNA']\n",
    "rrna_16s_mapping = {}\n",
    "for rrna_name in rrna:\n",
    "    for assembly_name in os.listdir(path):\n",
    "        assembly = os.path.join(path, assembly_name)\n",
    "#         print(get_16s_names(assembly))\n",
    "        rrna_16s_mapping[assembly_name] = get_16s(os.path.join(path, assembly), rrna_name)\n",
    "\n",
    "fasta_path = r'C:\\Users\\ALEX\\Desktop\\assemblies\\fasta_16s'\n",
    "\n",
    "with open(os.path.join(fasta_path, f\"Bacillus_16s.fasta\"), 'a+') as general:\n",
    "    for assembly in rrna_16s_mapping:\n",
    "        for i, seq_16s in enumerate(rrna_16s_mapping[assembly]):\n",
    "            assembly_name = assembly[:-5]\n",
    "            general.write(f'>{assembly_name}_{i}_Bacillus_16s\\n')\n",
    "            general.write(str(seq_16s) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
